pyrate/core/dem_error.py

```diff
@@ -0,0 +1,307 @@
+
+    # convert angles to radians
+    lon = lon / 180 * np.pi
+    lat = lat / 180 * np.pi
+    heading = heading / 180 * np.pi
+
+    # Earth radius at given latitude
+    re = np.sqrt(np.divide(np.square(a**2 * np.cos(lat)) + np.square(b**2 * np.sin(lat)), \
+                           np.square(a * np.cos(lat)) + np.square(b * np.sin(lat))))
+
+    # range measurement at pixel ij -> this requires the mli.par as input since rps would be different in the slc.par
+    range = near_range + rps*rg
+
+    # look angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    look_angle = np.arccos(np.divide(se**2 + np.square(range) - np.square(re), 2 * se * range))
+
+    # incidence angle at pixel ij -> law of cosines in "satellite - Earth centre - ground pixel" triangle
+    # see e.g. Section 2 in https://www.cs.uaf.edu/~olawlor/ref/asf/sar_equations_2006_08_17.pdf
+    incidence_angle = np.pi - np.arccos(np.divide(np.square(range) + np.square(re) - se**2, 2 * np.multiply(range, re)))
+    # for validation with GAMMA output only
+    incidence_angle_gamma = np.pi / 2 - incidence_angle
+    # maximum differences to the GAMMA-derived local incidence angles for Sentinel-1 test data are within +/-0.1 deg
+
+    # local azimuth angle at pixel ij using constant satellite heading angle and spherical approximations
+    epsilon = np.pi - look_angle - (np.pi - incidence_angle) # angle at the Earth's center between se and re
+    # azimuth of satellite look vector(for right-looking SAR)
+    sat_azi = heading + np.pi / 2
+    # the following equations are adapted from Section 4.4 (page 4-16) in EARTH-REFERENCED AIRCRAFT NAVIGATION AND
+    # SURVEILLANCE ANALYSIS (https://ntlrepository.blob.core.windows.net/lib/59000/59300/59358/DOT-VNTSC-FAA-16-12.pdf)
+    sat_lon = np.divide(np.arcsin(-(np.multiply(np.sin(epsilon), np.sin(sat_azi)))), np.cos(lat)) + lon # Eq. 103
+    temp = np.multiply(np.divide(np.cos(0.5 * (sat_azi + sat_lon - lon)), np.cos(0.5 * (sat_azi - sat_lon + lon))), \
+                       np.tan(0.5 * (np.pi / 2 + lat - epsilon))) # Eq. 104
+    sat_lat = -np.pi / 2 + 2 * np.arctan(temp)
+    # the above code could be improved by calculating satellite positions for each azimuth row
+    # from orbital state vectors given in .par file, using the following workflow:
+    # 1. read orbital state vectors and start/stop times from mli.par
+    # 2. for each pixel get the corresponding radar row (from matrix az)
+    # 3. get the corresponding radar time for that row (using linear interpolation)
+    # 4. calculate the satellite XYZ position for that time by interpolating the time and velocity state vectors
+
+    # calc azimuth angle using Vincenty's equations
+    azimuth_angle = np.empty(lat.shape) # pre-allocate 2D numpy array
+    azimuth_angle[:] = np.nan
```

Comment:
I think you can achieve the intent of these two lines in one (but I haven't tested my assumption):
`azimuth_angle = np.empty(lat.shape) * np.nan`

Reply:
That's right. Changed as suggested. This might however be slightly slower (see discussion here: https://stackoverflow.com/questions/1704823/create-numpy-matrix-filled-with-nans: "I like this one: a = numpy.empty((3, 3,)) * numpy.nan. It timed faster than fill but slower than the assignment method, but it is a oneliner!!")